rfc/0077-zone-aware-replica-reads.md

```diff
@@ -0,0 +1,183 @@
+network traffic when the server is being deployed across multiple availability
+zones.
+
+# Motivation
+
+Existing SDKs do not take into account node availability groups of the nodes in
+the cluster. In particular, when performing `getAllReplicas` operations, the SDK
+would send `GET (0x00)` request to active vBucket and `GET_REPLICA (0x83)` to
+each of the replicas. This approach will make reading from replicas expensive
+operations if nodes of the cluster deployed in different availability zones
+(AZ).
+
+# Use Cases
+
+## No Preference
+
+This is current behaviour, where the SDK does not take into account server
+groups.
+
+![figure 1: No Prefrence](figures/0077-case-1-no-preference.svg)
+
+## Local Only
+
+In this case, all reads will be done only from the local availability zone. It
+is the cheapest solution, although it reduces chances of getting data from
+replicas.
+
+![figure 2: Local Only](figures/0077-case-2-local-only.svg)
+
+## Local First
+
+This case is a compromise between cost and availability, which trades speed. The
+idea is that the SDK will partition set of `[active, replica1, ...]` nodes into
+two groups: local and non-local. Then it will perform requests to local only,
+and only when neither of nodes will return data, send requests to non-local
+group.
+
+![figure 3: Local First](figures/0077-case-3-local-first.svg)
+
+# Changes
+
+New enumeration should be defined for expressing different strategies:
+
+```
+enum ReadPreference {
+    NO_PREFERENCE,
+    LOCAL_ONLY,
+    LOCAL_FIRST,
+};
+```
+
+Each of the operation should get new option, with default value of
+`ReadPreference::NO_PREFERENCE`.
+
+```
+class GetAllReplicasOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class GetAnyReplicaOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class LookupInAllReplicasOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class LookupInAnyReplicaOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+In all failure cases, when the SDK cannot handle operation, it must return `102
+DocumentUnretrievable` with human-readable explanation whenever it is possible
+that explains the details.
+
+# Open Questions
+
+## Q1
+
+What the SDK should do if the local group does not have any replica configured
+for the key, but the User request `Local-Only` strategy? Consider the following
+cases:
+
+- number of replicas is 1, number of zones is 3. Neither active, nor replica is
+  in the local group.
+- number of replicas is 1, number of zones is 3. Only active node is in the
+  local group.
+- number of replicas is 1, number of zones is 3. Only replica node is in the
+  local group.
+
+Shall we warn users when locality restriction turns request into just GET,
```

Comment:
Yes, I raised the same concern in the one-pager - see there for Priya's thoughts.  

Warning seems a decent idea on the surface, but I worry it could get spammy.  I guess it only spams if there's a cluster misconfiguration, or in failover cases.

Reply:
Okay, no warnings. There will be good documentation on how to prepare the cluster.